Sources/BrowserServicesKit/SecureVault/AutofillDatabaseProvider.swift

```diff
@@ -98,96 +136,166 @@ public final class DefaultAutofillDatabaseProvider: GRDBSecureStorageDatabasePro
-                    throw SecureStorageError.duplicateRecord
-                } else {
-                    throw error
-                }
-            }
+    public func deleteSyncableCredentials(_ syncableCredentials: SecureVaultModels.SyncableCredentials, in database: Database) throws {
+        if let accountId = syncableCredentials.metadata.objectId {
+            try deleteWebsiteCredentialsForAccountId(accountId, in: database)
         }
+        try syncableCredentials.metadata.delete(database)
     }
 
-    func insertWebsiteCredentials(_ credentials: SecureVaultModels.WebsiteCredentials) throws -> Int64 {
+    public func deleteWebsiteCredentialsForAccountId(_ accountId: Int64) throws {
         try db.write {
-            do {
-                try credentials.account.insert($0)
-                let id = $0.lastInsertedRowID
-                try $0.execute(sql: """
-                    INSERT INTO
-                        \(SecureVaultModels.WebsiteCredentials.databaseTableName)
-                    (
-                        \(SecureVaultModels.WebsiteCredentials.Columns.id.name),
-                        \(SecureVaultModels.WebsiteCredentials.Columns.password.name)
-                    )
-                    VALUES (?, ?)
-                """, arguments: [id, credentials.password])
-                return id
-            } catch let error as DatabaseError {
-                if error.extendedResultCode == .SQLITE_CONSTRAINT_UNIQUE {
-                    throw SecureStorageError.duplicateRecord
-                } else {
-                    throw error
-                }
-            }
+            try deleteWebsiteCredentialsForAccountId(accountId, in: $0)
         }
     }
 
-    public func websiteCredentialsForAccountId(_ accountId: Int64) throws -> SecureVaultModels.WebsiteCredentials? {
-        return try db.read {
-            guard let account = try SecureVaultModels.WebsiteAccount.fetchOne($0, key: accountId) else {
-                return nil
-            }
+    private func deleteWebsiteCredentialsForAccountId(_ accountId: Int64, in database: Database) throws {
+        try updateSyncTimestamp(in: database, tableName: SecureVaultModels.SyncableCredentialsRecord.databaseTableName, objectId: accountId)
+        try database.execute(sql: """
+            DELETE FROM
+                \(SecureVaultModels.WebsiteAccount.databaseTableName)
+            WHERE
+                \(SecureVaultModels.WebsiteAccount.Columns.id.name) = ?
+            """, arguments: [accountId])
+    }
 
-            if let result = try Row.fetchOne($0,
-                                             sql: """
-                SELECT
-                    \(SecureVaultModels.WebsiteCredentials.Columns.password.name)
-                FROM
+    func updateWebsiteCredentials(in database: Database, _ credentials: SecureVaultModels.WebsiteCredentials, usingId id: Int64, timestamp: Date? = Date()) throws {
+        do {
```

Comment:
Maybe we should add `assert(database.isInsideTransaction)` here?

Reply:
Absolutely!